// @zeplin/sdk/dist/models/flow-board-shape-node.js
"use strict";

require("core-js/modules/es.object.define-property.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.transformJSONToFlowBoardShapeNode = exports.transformFlowBoardShapeNodeToJSON = void 0;

var _flowBoardNodeColor = require("./flow-board-node-color");

var _flowBoardPosition = require("./flow-board-position");

/* tslint:disable */

/* eslint-disable */

/**
 * Zeplin API
 * Access your resources in Zeplin
 *
 * Contact: support@zeplin.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var transformFlowBoardShapeNodeToJSON = function transformFlowBoardShapeNodeToJSON(value) {
  return {
    id: value.id,
    type: value.type,
    shape_type: value.shapeType,
    position: (0, _flowBoardPosition.transformFlowBoardPositionToJSON)(value.position),
    text: value.text,
    width: value.width,
    height: value.height,
    color: (0, _flowBoardNodeColor.transformFlowBoardNodeColorToJSON)(value.color)
  };
};

exports.transformFlowBoardShapeNodeToJSON = transformFlowBoardShapeNodeToJSON;

var transformJSONToFlowBoardShapeNode = function transformJSONToFlowBoardShapeNode(value) {
  return {
    id: value.id,
    type: value.type,
    shapeType: value.shape_type,
    position: (0, _flowBoardPosition.transformJSONToFlowBoardPosition)(value.position),
    text: value.text,
    width: value.width,
    height: value.height,
    color: (0, _flowBoardNodeColor.transformJSONToFlowBoardNodeColor)(value.color)
  };
};
/**
 * 
 * @export
 * @interface FlowBoardShapeNode
 */


exports.transformJSONToFlowBoardShapeNode = transformJSONToFlowBoardShapeNode;
