// @zeplin/sdk/dist/models/flow-board-node-color.js
"use strict";

require("core-js/modules/es.object.define-property.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.transformJSONToFlowBoardNodeColor = exports.transformFlowBoardNodeColorToJSON = void 0;

/* tslint:disable */

/* eslint-disable */

/**
 * Zeplin API
 * Access your resources in Zeplin
 *
 * Contact: support@zeplin.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var transformFlowBoardNodeColorToJSON = function transformFlowBoardNodeColorToJSON(value) {
  return {
    r: value.r,
    g: value.g,
    b: value.b,
    a: value.a
  };
};

exports.transformFlowBoardNodeColorToJSON = transformFlowBoardNodeColorToJSON;

var transformJSONToFlowBoardNodeColor = function transformJSONToFlowBoardNodeColor(value) {
  return {
    r: value.r,
    g: value.g,
    b: value.b,
    a: value.a
  };
};
/**
 * 
 * @export
 * @interface FlowBoardNodeColor
 */


exports.transformJSONToFlowBoardNodeColor = transformJSONToFlowBoardNodeColor;
