// @zeplin/cli/dist/src/tasks/ui/detect-project-type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.completed = exports.detecting = exports.skippedDetection = exports.initial = void 0;
function projectTypesText(ctx) {
    return ctx.projectTypes && ctx.projectTypes.length > 1
        ? `project types: ${ctx.projectTypes.map(pt => pt.type).join(", ")}`
        : `project type: ${ctx.projectTypes[0].type}`;
}
const initial = () => ({
    text: "Detect project type"
});
exports.initial = initial;
const skippedDetection = (ctx) => {
    var _a;
    return ({
        text: `Selected ${projectTypesText(ctx)}`,
        subtext: `You provided ${(_a = ctx.projectTypes) === null || _a === void 0 ? void 0 : _a.map(t => `--type=${t.type}`).join(" ")}`
    });
};
exports.skippedDetection = skippedDetection;
const detecting = () => ({
    text: "Detecting project type..."
});
exports.detecting = detecting;
const completed = (ctx) => {
    if (!ctx.projectTypes || ctx.projectTypes.length === 0) {
        return {
            text: "Could not detect project type",
            subtext: "You may use --type parameter to enforce project type"
        };
    }
    return { text: `Detected ${projectTypesText(ctx)}` };
};
exports.completed = completed;
