// @rtk-query/codegen-openapi/lib/codegen.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateEndpointDefinition = exports.generateCreateApiCall = exports.generateImportNode = exports.generateObjectProperties = void 0;
const typescript_1 = __importDefault(require("typescript"));
const factory_1 = require("./utils/factory");
const defaultEndpointBuilder = factory_1.factory.createIdentifier('build');
function generateObjectProperties(obj) {
    return Object.entries(obj).map(([k, v]) => factory_1.factory.createPropertyAssignment(factory_1.factory.createIdentifier(k), v));
}
exports.generateObjectProperties = generateObjectProperties;
function generateImportNode(pkg, namedImports, defaultImportName) {
    return factory_1.factory.createImportDeclaration(undefined, undefined, factory_1.factory.createImportClause(false, defaultImportName !== undefined ? factory_1.factory.createIdentifier(defaultImportName) : undefined, factory_1.factory.createNamedImports(Object.entries(namedImports).map(([propertyName, name]) => factory_1.factory.createImportSpecifier(name === propertyName ? undefined : factory_1.factory.createIdentifier(propertyName), factory_1.factory.createIdentifier(name))))), factory_1.factory.createStringLiteral(pkg));
}
exports.generateImportNode = generateImportNode;
function generateCreateApiCall({ endpointBuilder = defaultEndpointBuilder, endpointDefinitions, }) {
    return factory_1.factory.createVariableStatement(undefined, factory_1.factory.createVariableDeclarationList([
        factory_1.factory.createVariableDeclaration(factory_1.factory.createIdentifier('injectedRtkApi'), undefined, undefined, factory_1.factory.createCallExpression(factory_1.factory.createPropertyAccessExpression(factory_1.factory.createIdentifier('api'), factory_1.factory.createIdentifier('injectEndpoints')), undefined, [
            factory_1.factory.createObjectLiteralExpression(generateObjectProperties({
                endpoints: factory_1.factory.createArrowFunction(undefined, undefined, [
                    factory_1.factory.createParameterDeclaration(undefined, undefined, undefined, endpointBuilder, undefined, undefined, undefined),
                ], undefined, factory_1.factory.createToken(typescript_1.default.SyntaxKind.EqualsGreaterThanToken), factory_1.factory.createParenthesizedExpression(endpointDefinitions)),
                overrideExisting: factory_1.factory.createFalse(),
            }), true),
        ])),
    ], typescript_1.default.NodeFlags.Const));
}
exports.generateCreateApiCall = generateCreateApiCall;
function generateEndpointDefinition({ operationName, type, Response, QueryArg, queryFn, endpointBuilder = defaultEndpointBuilder, extraEndpointsProps, }) {
    return factory_1.factory.createPropertyAssignment(factory_1.factory.createIdentifier(operationName), factory_1.factory.createCallExpression(factory_1.factory.createPropertyAccessExpression(endpointBuilder, factory_1.factory.createIdentifier(type)), [Response, QueryArg], [
        factory_1.factory.createObjectLiteralExpression(generateObjectProperties({ query: queryFn, ...extraEndpointsProps }), true),
    ]));
}
exports.generateEndpointDefinition = generateEndpointDefinition;
